### PR TITLE
Add LogSlot for logging blocked requests and refactor logger

### PR DESCRIPTION
**Pull request by sczyh30**
Signed-off-by: Eric Zhao <sczyh16@gmail.com>

### Describe what this PR does / why we need it

Add `LogSlot` to log blocked requests.

### Does this pull request fix one issue?

NONE

### Describe how you did it

- Add a `BlockLogTask` to roll blocked log per second for each `resource|cause` pair.
- Add `LogSlot` to log blocking requests, which is useful to investigate "blocked" events. The `LogSlot` is placed at the last position of the default slot chain.
- Preserve warning log when resource size exceeds the threshold (6000).

### Describe how to verify it

Run the demo and see the generated log in `~/logs/csp/sentinel-block.log`.

### Special notes for reviews

NONE

### CODE REVIEW

File: sentinel-core/log/block/block_log_task.cc [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+namespace Log {
+
+BlockLogTask::BlockLogTask() {
+  auto filename = LogBase::GetLogBaseDir() + kBlockLogFilename;
+  try {
+    logger_ = spdlog::rotating_logger_mt(kBlockLoggerName, filename,
+                                         kDefaultBlockLogMaxSize, 3);
+    logger_->set_pattern("%v");
+  } catch (const spdlog::spdlog_ex& ex) {
+    std::cerr << "Sentinel block log initialization failed: " << ex.what()
```
Comment (zyfjeff) [inline sentinel-core/log/block/block_log_task.cc]: If the logger to create failure, whether there is a need to start a thread?

File: sentinel-core/log/block/block_log_task.cc [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+          int64_t cur_time = TimeUtils::CurrentTimeMillis().count();
+          auto time_str = absl::FormatTime("%Y-%m-%d %H:%M:%S",
+                                           absl::FromUnixMillis(cur_time),
+                                           absl::LocalTimeZone());
+          // format: time|resource|exception
+          logger_->info("{}|{}", time_str, e.first);
+          e.second.last_write_ = cur_time;
+        }
+      }
+      logger_->flush();
```
Comment (zyfjeff) [inline sentinel-core/log/block/block_log_task.cc]: clear map?

File: sentinel-core/log/block/block_log_task.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+  void Start();
+  void Stop();
+
+  void Log(const std::string& resource, const std::string& cause);
+
+  bool started() const { return started_; }
+
+ private:
+  bool started_{false};
```
Comment (zyfjeff) [inline sentinel-core/log/block/block_log_task.h]: replace atomic?

File: sentinel-core/log/block/block_log_task.h [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+namespace Sentinel {
+namespace Log {
+
+static constexpr const char* kBlockLoggerName = "sentinel_block_logger";
+static constexpr const char* kBlockLogFilename = "sentinel-block.log";
+static constexpr uint32_t kDefaultBlockLogMaxSize = 1024 * 1024 * 300;
+
+struct BlockLogRecord {
+ public:
```
Comment (zyfjeff) [inline sentinel-core/log/block/block_log_task.h]: default public

File: sentinel-core/log/block/block_log_task.cc (hunk 1 above)
Comment (sczyh30) [reply sentinel-core/log/block/block_log_task.cc]: I've added checking in `Start()` method. If the logger is absent, the daemon thread won't start.

File: sentinel-core/log/block/block_log_task.cc [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+                                             absl::LocalTimeZone());
+            // format: time|resource|exception
+            logger_->info("{}|{}", time_str, e.first);
+            e.second.last_write_ = cur_time;
+          }
+        }
+      }
+      logger_->flush();
+      {
+        absl::WriterMutexLock lck(&mtx_);
```
Comment (zyfjeff) [inline sentinel-core/log/block/block_log_task.cc]: After the `logger_->flush()` calls, if the added an element in the map, cleaned here will lead to loss of data.

File: sentinel-core/log/block/block_log_task.cc [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    std::thread daemon_task(&BlockLogTask::LoopWriteBlockLog, this);
+    daemon_task.detach();
+  }
+}
+
+void BlockLogTask::Stop() { started_.store(false); }
+
+void BlockLogTask::Log(const std::string& resource, const std::string& cause) {
+  auto key = absl::StrFormat("%s|%s", resource, cause);
+  mtx_.ReaderLock();
```
Comment (zyfjeff) [inline sentinel-core/log/block/block_log_task.cc]: use lock guard, and refactor this code
```
{
  absl::ReaderMutexLock lck(&mtx_);
  auto it = map_.find(key);
  if (it != map_.end()) {
     it->second.last_block_ = TimeUtils::CurrentTimeMillis().count();
     return;
  }
}
    absl::WriterMutexLock lck(&mtx_);
    map_.emplace(std::make_pair(
        key, BlockLogRecord{0, TimeUtils::CurrentTimeMillis().count()}));
```

File: sentinel-core/log/block/block_log_task.cc (hunk 5 above)
Comment (sczyh30) [reply sentinel-core/log/block/block_log_task.cc]: Yes, there might be data loss here, but it's okay. I'll use a writer lock to wrap the entire scope of `LoopWriteBlockLog()`.

File: sentinel-core/log/block/block_log_task.cc (hunk 6 above)
Comment (sczyh30) [reply sentinel-core/log/block/block_log_task.cc]: Done. Please check.

**Comment by zyfjeff** [comment]
Please add some basic test cases

**Comment by sczyh30** [comment]
> Please add some basic test cases @sczyh30

I've added basic test cases and refactored the new logger (from master). Please check.

**Comment by sczyh30** [comment]
I've reformatted the code using `format.sh` but the CI still failed...
